### PR TITLE
txmempool: Make entry time type-safe (std::chrono)

### PR DESCRIPTION
**Pull request by maflcko**
This changes the type of the entry time of txs into the mempool from `int64_t` to `std::chrono::seconds`.

The benefits:
* Documents the type for developers
* Type violations result in compile errors
* After compilation, the two are equivalent (at no run time cost)

### CODE REVIEW

**Comment by emilengler** [comment]
Concep ACK,
chrono is definitely better for time managment

**Comment by ajtowns** [comment]
ConceptACK

I think `LimitMemPoolSize()` should have `std::chrono::seconds age` instead of `unsigned long age`, which means you can just pass in `std::chrono::hours{gArgs.GetArg(...)}`; makes it a bit tidier.

Having `int64_t as_seconds(std::chrono::seconds t) { return t.count(); }` seems like it would be better than using `int64_t{xxx.count()}` fwiw; that way if we change xxx's type either the conversion happens automatically if there's no loss of precision, or you get a compile time error.

**Comment by MarcoFalke** [comment]
Thx for the review. Took all suggestions by you.

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -418,7 +418,7 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
     info.pushKV("weight", (int)e.GetTxWeight());
     info.pushKV("fee", ValueFromAmount(e.GetFee()));
     info.pushKV("modifiedfee", ValueFromAmount(e.GetModifiedFee()));
-    info.pushKV("time", e.GetTime());
+    info.pushKV("time", int64_t{count_seconds(e.GetTime())});
```
Comment (ajtowns) [inline src/rpc/blockchain.cpp]: No need for the `int64_t{}` here?

Comment (MarcoFalke) [reply src/rpc/blockchain.cpp]: Done

**Comment by ajtowns** [comment]
utACK faec689bed7a5b66e2a7675853d10205b933cec8

File: src/net_processing.cpp [hunk 2]
```diff
@@ -1541,11 +1541,11 @@ void static ProcessGetData(CNode* pfrom, const CChainParams& chainparams, CConnm
             if (mi != mapRelay.end()) {
                 connman->PushMessage(pfrom, msgMaker.Make(nSendFlags, NetMsgType::TX, *mi->second));
                 push = true;
-            } else if (pfrom->m_tx_relay->timeLastMempoolReq) {
+            } else if (pfrom->m_tx_relay->m_last_mempool_req.load().count()) {
```
Comment (laanwj) [inline src/net_processing.cpp]: why not use `count_seconds` here, if it is preferred to inline `.count()` (according to the comment)?

Comment (MarcoFalke) [reply src/net_processing.cpp]: The member uses a magic value of 0 to indicate that the mempool was never requested. So this check isn't about time (in a specific type like seconds), but more about checking that any value exists.

Comment (laanwj) [reply src/net_processing.cpp]: Hmm, okay, I guess we'd really want a `optional` here some day instead of a magic value, but no need to do that here.

**Comment by laanwj** [comment]
ACK faec689bed7a5b66e2a7675853d10205b933cec8

File: src/net_processing.cpp (hunk 2 above)
Comment (sdaftuar) [reply src/net_processing.cpp]: Sorry to chime in late -- would it make sense to use the `std::chrono::duration::zero` value here?  I found `count()` to be unintuitive when I was looking at related code just now (but this `std::chrono` stuff is new to me so I'm still figuring out how to use it).

Comment (MarcoFalke) [reply src/net_processing.cpp]: Comparing it to zero would also work, but still be odd, as the code is written as if the member were an `optional`.

I think we can just stop treating it as `optional` (always assume the last mempool request time for peers was in year 1970). But that can be done in a follow up pull request.

**Pull request merged**